### DISCUSSION TITLE
Questions About Distributed Training

### DISCUSSION THREAD

**Discussion by Endeavour10020** [open-mmlab/mmediting#1278] [How-to]
Hi, guys. I meet some issues when trying to use multi-gpus in distribution mode. I found it not faster than a single GPU. Further, I found more GPU take longer training time. So, I make an experiment to verify this. Take the config of `configs/synthesizers/pix2pix/pix2pix_vanilla_unet_bn_a2b_1x1_219200_maps.py` for example, the estimated time of the training process is 2h, 6h, 8h for 1GPU, 2GPU, 8GPU respectively.
Is this a normal phenomenon (I guess not)? How to use distributed training correctly to speed up my experiment? 
I list the commands that I used below.

1gpu: `python ./tools/train.py configs/synthesizers/pix2pix/pix2pix_vanilla_unet_bn_a2b_1x1_219200_maps.py`

2gpus: `bash ./tools/dist_train.sh configs/synthesizers/pix2pix/pix2pix_vanilla_unet_bn_a2b_1x1_219200_maps.py 2`

8gpus: `bash ./tools/dist_train.sh configs/synthesizers/pix2pix/pix2pix_vanilla_unet_bn_a2b_1x1_219200_maps.py 8`

1gpu:
![image](https://user-images.githubusercontent.com/55418760/166879449-ff97cb33-2ecd-4eb4-a332-8e1154e04601.png)

2gpus:
![image](https://user-images.githubusercontent.com/55418760/166879657-f79ec41c-744b-4575-a553-eefb159e178b.png)

8gpus:
![image](https://user-images.githubusercontent.com/55418760/166879698-d08ad3bc-5465-4aee-b674-a1e66f5de1c6.png)

So, how to solve this problem?

**Comment by wangruohui** [comment]
Hello, 

Not quite normal. Even though you don't decrease the batch size, training time should be compareble when using different numbers of GPUs. 

Are you using 8 GPU in a single node or multiple nodes? Could you please upload the result of `nvidia-smi topo -m`?

Moreover, image translation models have been moved to [MMGeneration](https://github.com/open-mmlab/mmgeneration), with a new design. You can try MMGeneration.

**Comment by Endeavour10020** [comment]
Here the result of `nvidia-smi topo -m`
![image](https://user-images.githubusercontent.com/55418760/166881497-d2dcfd09-9ee0-4afe-93f4-191a050761c4.png)

The server is equipped with 8 2080Ti Gpus. I use them all in a single node.

I am familiar with some other repos in openmmlab, e.g. MMaction2. The distribution mode woks well in that repo. 

By the way, MMediting uses iterations, not epochs to control the whole training process. I am not sure whether it is the latent reason.

**Comment by Endeavour10020** [comment]
@wangruohui 
Hi ruohui,
Maybe you can clone the latest MMediting repo, and try to train the pix2pix model with the distribution command.

**Comment by wangruohui** [comment]
It should have nothing to do with iter or epoch. It might have something to do with `DistributedDataParallelWrapper`.

Have you tried MMGeneration?

**Comment by Endeavour10020** [comment]
Haven't tried yet. But if the problem could not be fixed, I will turn to MMGeneration. Does MMediting code works well on your computer under distribution mode?

**Comment by wangruohui** [comment]
I don't have a prepared environment at hand. Maybe you need to wait for some time if you want me to have a try. 

But actually, since these two models are supported by MMGeneration for nearly a year, we are considering deprecating them. So I still suggest you switch to MMGeneration anyway :>

**Comment by Endeavour10020** [comment]
Ok, thank you for your advice!

**Comment by Endeavour10020** [comment]
@wangruohui 
Hi ruohui, 
I found the same problem in MMGeneration. 8gpus takes longer training time than single gpu while using the same config file. why?

**Comment by wangruohui** [comment]
I tried, but my task is still pending as I borrow some machines from others :(

Previously I thought the problem lies in your hardware because SYS is the slowest connection within a node. multi GPU running would be problematic. But you said that on action2 it works well, so maybe you need to wait until my tasks get run.

**Comment by wangruohui** [comment]
OK, I just got some data. 
1 GPU 2.5h
2 GPU 4.2h

**Comment by Endeavour10020** [comment]
So, is this a bug? And I found the training time increases with the whole batch size (samples_per_gpu*gpu_number). 
Its weird, right?

**Comment by wangruohui** [comment]
probably not a bug, just lack of performance optimization. Smaller model suffer usually more communication overhead.
You can try using torch profiler to profile the runtime if you hope dig into this problem.

**Comment by sciart17** [comment]
Hi, guys. Have the problem been solved? I met the same issues as mentioned above. More GPU take longer training time and bigger sample_per_gpu take longer training time as well.
